src/platform/input/game-pads.js

```diff
@@ -68,62 +109,645 @@ const MAPS = {
+    }
+
+    /**
+     * Update the existing GamePad Instance.
+     *
+     * @param {Gamepad} gamepad - The original Gamepad API gamepad.
+     * @ignore
+     */
+    update(gamepad) {
+        this.pad = gamepad;
+
+        const buttons = this._buttons;
+        for (let i = 0, l = buttons.length; i < l; i++) {
+            const button = buttons[i];
+
+            if (button)
+                button.update(gamepad.buttons[i]);
+        }
+
+        return this;
+    }
+
+    /**
+     * Update the previous values for buttons and axes.
+     *
+     * @ignore
+     */
+    updatePrevious() {
+        const buttons = this._buttons;
+        for (let i = 0, m = buttons.length; i < m; i++) {
+            buttons[i].updatePrevious();
+        }
+
+        // Store previous values for axes for dual buttons.
+        const previousAxes = this._previousAxes;
+        previousAxes.length = 0;
+        previousAxes.push(...this.pad.axes);
+    }
+
+    /**
+     * The values from analog axes present on the GamePad. Values are between -1 and 1.
+     *
+     * @type {number[]}
+     */
+    get axes() {
+        return this.map.axes ? this.map.axes.map(a => this.pad.axes[MAPS_INDEXES.axes[a]] || 0) : [];
+    }
+
+    /**
+     * The buttons present on the GamePad. Some buttons may be null.
+     *
+     * @type {GamePadButton[]}
+     */
+    get buttons() {
+        return this.map.buttons ? this.map.buttons.map(b => this._getButton(b)) : [];
+    }
+
+    /**
+     * Make the gamepad vibrate.
+     *
+     * @param {number} intensity - Intensity for the vibration in the range 0 to 1.
+     * @param {number} duration - Duration for the vibration in milliseconds.
+     * @param {object} [options] - Options for special vibration pattern.
+     * @param {string} [options.type] - Type of pattern. Available types are "dual-rumble" and "vibration". Defaults to "vibration".
+     * @param {number} [options.startDelay] - Delay before the pattern starts, in milliseconds. Defaults to 0.
+     * @param {number} [options.strongMagnitude] - Intensity for strong actuators in the range 0 to 1. Defaults to intensity.
+     * @param {number} [options.weakMagnitude] - Intensity for weak actuators in the range 0 to 1. Defaults to intensity.
+     * @returns {Promise<boolean>} Return a Promise resulting in true if the pulse was successfully completed.
+     */
+    async pulse(intensity, duration, options) {
+        const actuators = this.pad.vibrationActuator ? [this.pad.vibrationActuator] : this.pad.hapticActuators || [];
+
+        if (actuators.length) {
+            const startDelay = options?.startDelay ?? 0;
+            const strongMagnitude = options?.strongMagnitude ?? intensity;
+            const weakMagnitude = options?.weakMagnitude ?? intensity;
+
+            const results = await Promise.all(
+                actuators.map(async (actuator) => {
+                    if (!actuator) {
+                        return true;
+                    }
```

Comment:
Is this check really needed? My test on Linux/Firefox doesn't show any `[null]` array or something. And if this is really needed, I guess it should return `false` instead?

![image](https://user-images.githubusercontent.com/5236548/219954735-0257e4be-5e21-4769-8026-365ae9500446.png)

Reply:
There is no clear documentation on whether or not it's nullable, so I went better safe than sorry. And no if there is none it will return an error when checking if the functions exist inside.